README
== About ==

The 'Variables' extension for MediaWiki introduces parser functions for defining page-scoped
variables within wiki pages.

* Website: https://www.mediawiki.org/wiki/Extension:Variables
* Authors: Rob Adams, Tom Hempel, Xiloynaha and Daniel Werner
* License: Public domain


== Installation ==

Once you have downloaded the code, place the 'Variables' directory within your
MediaWiki 'extensions' directory. Then add the following code to your
[[Manual:LocalSettings.php|LocalSettings.php]] file:

 # Variables
 require_once( "$IP/extensions/Variables/Variables.php" );



RELEASE-NOTES
=== git era Changelog ===

Please note that the git era changelog isn't complete regarding internal adjustments.
Only things that are major architecture change or might break compatibility with MediaWiki
versions and other extensions is collected here.

; TBC: v2.6.0, by MGChecker, Func and …
* Fix further template caching issues
** Template caching could mess up resetting Variables via repeated template calls.
   This won't happen anymore, as the parser frame is marked as volatile by the extension.
** This new behavior can be deactivated by setting '$egVariablesAreVolatile' to false,
   however be aware that this is a temporary option for transition and will be removed in
   the future.
* Use the default StripState for '#var_final' (T221080, T250963)

; 2019-07-11: v2.5.2, by MGChecker
* Removal of PHP entry point, use wfLoadExtension( 'Variables' ) instead

; 2019-03-09: v2.5.1, by MGChecker
* Moving hook functions to separate VariablesHooks class (T202412)
* Moving code files into includes/ directory
* Renaming files to match class name

; 2018-08-30: v2.5.0, by MGChecker
* Fix '#varexists' parsing (T202679)
** if and else value are no longer expanded if not needed, thus variable definitions can be
   done conditionally and making the parser faster

; 2018-08-24: v2.4.1, by MGChecker
* Class attributes of ExtVariables changed to private, as it has been annotated before.

; 2018-04-20: v2.4.0, by MGChecker
* Fix longstanding template caching issues (T191574)
** Template caching could mess up definition of Variables depending on other Variables in
   templates, if they are called multiple times. This won't happen anymore, as the parser
   frame is marked as volatile by the extension.
** This new behavior can be deactivated by setting '$egVariablesAreVolatile' to false,
   however be aware that this is a temporary option for transition and will be removed in
   the future.

; 2018-01-23: v2.3.0 by MGChecker
* Conversion to Extension registration (T173625)
* ExtVariables::VERSION is deprecated, use Extension registration instead [1]
* Drop support for MediaWiki before 1.29
* Remove Variables.settings.php file, as supersed by Extension registration

; 2017-09-25: (trunk) by Leonid Verhovskij
* Add composer.json for easy autoloading of required librariers and the extension itself

; 2017-07-27: v2.2.0 by Karsten Hoffmeyer
* Removal of backwards compatibility code for old MediaWiki versions
** Removal of i18n compatibility shim
** Removal of InternalParseBeforeLinks hook

; 2016-03-07: (trunk) by Antoine Musso
* Removal of the unique marker passed to StripState
* Drop support support for MediaWiki before 1.26

; 2014-04-01: v2.1.0 by Siebrand Mazeland
* Migration from PHP i18n to JSON i18n
* Addition of PHP i18n compatibility shim

; 2012-06-24: v2.0.1 by Daniel Werner
* Using 'InternalParseBeforeSanitize' hook instead of 'InternalParseBeforeLinks' for MW 1.20
* Put into gerrit.wikimedia.org git repository.

=== svn era Changelog ===

; 2010-11-16: v2.0 by Daniel Werner
* Version 2.0 almost is a complete rewrite of the extension, just the idea remains the
  same. It's the attempt to get rid of several bugs caused by the fact that MediaWiki
  is using several Parser objects. Therefore in v2 each Parser has its own Variables
  store, so nothing will get mixed up anymore. Full compatibility is given, except perhaps
  in cases where bugs were used intentionally.

** New features and bugfixes:
*** Inclusion of special pages in the middle of the page won't reset all defined
    variables anymore.
*** Variables should now be compatible with all other extensions, except for those still
    using Parser::parse() recursively in any way (which should never be done!).
*** For MW 1.12 and later, '#var' default value no longer gets expanded when not needed.
*** Experimental new function '#var_final' which allows to insert the variables final
    (last) value after page processing is almost through.
*** Global configuration variable '$egVariablesDisabledFunctions' added.

** Internal changes:
*** Parser class member $mExtVariables now contains an instance of ExtVariables where
    only variables for that parser are getting stored. They won't be deleted by other
    Parser actions anymore (e.g. special page inclusion doesn't reset variables anymore)
*** ExtVariables class now has public functions which should be used by other extensions
    for getting and setting variables information.
*** Removed global '$wgExtVariables' variable.
*** 'Variables_Settings.php' file for configuration settings added.

** Others:
*** Put under 'ISC License' (public domain before).
*** In case you are using Extension:Loops, you should update it as well to remain compatibility.


; 2011-11-14: v1.3.1.1 (re-pack of 1.3.1 tagged version for svn)
* Although 1.4 is released already and 2.0 in the trunk, ensure one last version
  compatible with MW < 1.12 still is available
* Some comments cleaned, 'RELEASE-NOTES' and 'README' files added.


; 2011-11-13: v1.4 by Daniel Werner
* Cleanup for use with more current MW versions:
** 'ParserFirstCallInit' hook in use and no more global extension functions.
** State of the Art internationalization files added.
* Dropped support for MW before 1.12
* 'ExtVariables::VARIABLES' constant with version info added.
* Put into mediawiki.org svn, 'RELEASE-NOTES' and 'README' files added.

=== Pre svn Changelog ===

The following pre-svn changelog was composed by Daniel Werner in the hope it might
give an almost complete overview of all major releases of 'Variables' extension.
All changes can still be retrace at [2]

; 2010-07-20: v1.3 by Daniel Werner
* Removed critical bug. Some kind of "Superglobal" variables. In some cases values
  were passed from one page to another page during page imports and job queue jobs.

; 2009-03-28: v1.2 by Daniel Werner
* '#varexists' function introduced
* parameter for default value for '#var' function in case the variable doesn't exist

tests/phan/config.php
<?php

$config = require __DIR__ . '/../../vendor/mediawiki/mediawiki-phan-config/src/config.php';
// Due to creation of Parser::$mExtVariables property
$config['suppress_issue_types'][] = 'PhanUndeclaredProperty';

return $config;


Variables.i18n.php
<?php
/**
 * This is a backwards-compatibility shim, generated by:
 * https://git.wikimedia.org/blob/mediawiki%2Fcore.git/HEAD/maintenance%2FgenerateJsonI18n.php
 *
 * Beginning with MediaWiki 1.23, translation strings are stored in json files,
 * and the EXTENSION.i18n.php file only exists to provide compatibility with
 * older releases of MediaWiki. For more information about this migration, see:
 * https://www.mediawiki.org/wiki/Requests_for_comment/Localisation_format
 *
 * This shim maintains compatibility back to MediaWiki 1.17.
 */
$messages = array();
if ( !function_exists( 'wfJsonI18nShim46fcc9ed914bb69c' ) ) {
	function wfJsonI18nShim46fcc9ed914bb69c( $cache, $code, &$cachedData ) {
		$codeSequence = array_merge( array( $code ), $cachedData['fallbackSequence'] );
		foreach ( $codeSequence as $csCode ) {
			$fileName = dirname( __FILE__ ) . "/i18n/$csCode.json";
			if ( is_readable( $fileName ) ) {
				$data = FormatJson::decode( file_get_contents( $fileName ), true );
				foreach ( array_keys( $data ) as $key ) {
					if ( $key === '' || $key[0] === '@' ) {
						unset( $data[$key] );
					}
				}
				$cachedData['messages'] = array_merge( $data, $cachedData['messages'] );
			}

			$cachedData['deps'][] = new FileDependency( $fileName );
		}
		return true;
	}

	$GLOBALS['wgHooks']['LocalisationCacheRecache'][] = 'wfJsonI18nShim46fcc9ed914bb69c';
}


ExtVariables.class.php
<?php

/**
 * Extension class with basic extension information. This class serves as static
 * class with the static parser functions but also as variables store instance
 * as object assigned to a Parser object.
 */
class ExtVariables {

	/**
	 * Version of the 'Variables' extension.
	 * Using this constant is deprecated, please use the data in extension.json instead.
	 *
	 * @deprecated since 2.3.0
	 * @since 1.4
	 *
	 * @var string
	 */
	const VERSION = '2.5.0';

	/**
	 * Internal store for variable values
	 *
	 * @var string[]
	 */
	private $mVariables = [];

	/**
	 * Array with all names of variables requested by '#var_final'. Key of the values is the
	 * stripSateId of the strip-item placed where the final var should appear.
	 *
	 * @since 2.0
	 *
	 * @var string[]
	 */
	private $mFinalizedVars = [];

	/**
	 * Variables extensions own private StripState manager to manage 'finalizedvar' placeholders
	 * and their replacement with the final var value or a defined default.
	 *
	 * @since 2.0
	 *
	 * @var StripState
	 */
	private $mFinalizedVarsStripState;

	/**
	 * Sets up parser functions
	 *
	 * @since 1.4
	 *
	 * @param Parser $parser
	 */
	public static function init( Parser $parser ) {
		$self = new self();

		/*
		 * store for variables per parser object. This will solve several bugs related to
		 * 'ParserClearState' hook clearing all variables early in combination with certain
		 * other extensions. (since v2.0)
		 */
		$parser->mExtVariables = $self;

		$self->initFunction( $parser, 'var', Parser::SFH_OBJECT_ARGS );
		$self->initFunction( $parser, 'varexists', Parser::SFH_OBJECT_ARGS );
		$self->initFunction( $parser, 'var_final' );
		$self->initFunction( $parser, 'vardefine' );
		$self->initFunction( $parser, 'vardefineecho' );
	}

	/**
	 * Does the actual setup after making sure the functions aren't disabled
	 *
	 * @param Parser $parser
	 * @param string $name The name of the parser function
	 * @param int $flags Some configuration options, see also definition in Parser.php
	 */
	private function initFunction( Parser $parser, $name, $flags = 0 ) {
		// register function only if not disabled by configuration:
		global $egVariablesDisabledFunctions;

		if ( !in_array( $name, $egVariablesDisabledFunctions ) ) {
			$parser->setFunctionHook(
				$name,
				[ $this, $flags & Parser::SFH_OBJECT_ARGS ? "pfObj_$name" : "pf_$name" ],
				$flags
			);
		}
	}

	// Parser Functions

	/**
	 * Sets up #vardefine parser function to save variable values internally.
	 * The parameters of the parser function correspond with the parameters of this function.
	 *
	 * @param Parser $parser The parser instance these variables are bound to
	 * @param string $varName The name of the variable
	 * @param string $value The value of the defined variable
	 *
	 * @return string '' This parser function has no output
	 */
	public function pf_vardefine( Parser $parser, $varName = '', $value = '' ) {
		$this->get( $parser )->setVarValue( $varName, $value );
		return '';
	}

	/**
	 * Sets up #vardefineecho parser function to save variable values internally.
	 * The parameters of the parser function correspond with the parameters of this function.
	 *
	 * @param Parser $parser The parser instance these variables are bound to
	 * @param string $varName The name of the variable
	 * @param string $value The value of the defined variable
	 *
	 * @return string The value assigned to the variable
	 */
	public function pf_vardefineecho( Parser $parser, $varName = '', $value = '' ) {
		$this->get( $parser )->setVarValue( $varName, $value );
		return $value;
	}

	/**
	 * Sets up #varexists parser function to check if a variable was ever defined on a page.
	 * Even if an empty string is assigned to an variable, it still exists for this function.
	 * The parameters of the parser function correspond with the content of the $args param.
	 *
	 * @param Parser $parser The parser instance these variables are bound to
	 * @param PPFrame $frame The current frame
	 * @param array $args The arguments of the parser function
	 *
	 * @return string the content of the second or third parameter
	 * If none are provided, 1 or empty string by default
	 */
	public function pfObj_varexists( Parser $parser, PPFrame $frame, array $args ) {
		// first argument expanded already but lets do this anyway
		$varName = isset( $args[0] ) ? trim( $frame->expand( $args[0] ) ) : '';

		// this prevents issues due to template caching,
		// templates using variables are reparsed every call.
		global $egVariablesAreVolatile;
		if ( $egVariablesAreVolatile ) {
			$frame->setVolatile();
		}

		// if you expand these arguments earlier, you parse wikitext you discard later on.
		// doing so would lead to unexpected effects and decrease performance.
		if ( $this->get( $parser )->varExists( $varName ) ) {
			$exists = isset( $args[1] ) ? trim( $frame->expand( $args[1] ) ) : '1';
			return $exists;
		} else {
			$noexists = isset( $args[2] ) ? trim( $frame->expand( $args[2] ) ) : '';
			return $noexists;
		}
	}

	/**
	 * Sets up #var parser function to return the value of a variable.
	 *
	 * @param Parser $parser The parser instance these variables are bound to
	 * @param PPFrame $frame The current frame
	 * @param array $args The arguments of the parser function
	 *
	 * @return string the value assigned to the variable
	 */
	public function pfObj_var( Parser $parser, PPFrame $frame, array $args ) {
		// first argument expanded already but lets do this anyway
		$varName = trim( $frame->expand( $args[0] ) );
		$varVal = $this->get( $parser )->getVarValue( $varName, null );

		// this prevents issues due to template caching,
		// templates using variables are reparsed every call
		global $egVariablesAreVolatile;
		if ( $egVariablesAreVolatile ) {
			$frame->setVolatile();
		}

		// default applies if var doesn't exist but also in case it is an empty string!
		if ( $varVal === null || $varVal === '' ) {
			// only expand argument when needed:
			$defaultVal = isset( $args[1] ) ? trim( $frame->expand( $args[1] ) ) : '';
			return $defaultVal;
		}
		return $varVal;
	}

	/**
	 * Sets up #var_final parser function to call the final value of an variable
	 * after parsing the page is done.
	 * The parameters of the parser function correspond with the parameters of this function.
	 *
	 * @param Parser $parser The parser instance these variables are bound to
	 * @param string $varName The name of the variable
	 * @param string $defaultVal The output if no such variable is defined
	 *
	 * @return string
	 */
	public function pf_var_final( Parser $parser, $varName, $defaultVal = '' ) {
		return $this->get( $parser )->requestFinalizedVar( $parser, $varName, $defaultVal );
	}

	// Used Hooks

	/**
	 * Used for '#var_final' parser function to insert the final variable values.
	 * @see http://www.mediawiki.org/wiki/Manual:Hooks/InternalParseBeforeSanitize
	 *
	 * @since 2.0.1
	 *
	 * @param Parser $parser
	 * @param string &$text The text to parse
	 */
	public static function onInternalParseBeforeSanitize( Parser $parser, &$text ) {
		$varStore = self::get( $parser );

		// only do this if '#var_final' was used
		if ( $varStore->mFinalizedVarsStripState === null ) {
			return;
		}

		/*
		 * all vars are final now, check whether requested vars can be inserted for 'finalizedvar' or
		 * if the default has to be inserted. In any case, adjust the strip item value
		 */
		foreach ( $varStore->mFinalizedVars as $stripStateId => $varName ) {

			$varVal = $varStore->getVarValue( $varName, '' );
			if ( $varVal !== '' ) {
				// replace strip item value with final variables value or registered default:
				$varStore->stripStatePair( $stripStateId, $varVal );
			}
		}

		/**
		 * Unstrip all '#var_final' strip-markers with their final '#var' or default values.
		 * This HAS to be done here and can't be done through the normal unstrip process of MW.
		 * This is because the default value as well as the variables value still have to be
		 * rendered properly since they may contain links or even category links.
		 * On the other hand, they can't be parsed with Parser::recursiveTagParse() since this
		 * would parse wiki templates and functions which are intended as normal text, kind of
		 * similar to returning a parser functions value with 'noparse' => true.
		 * Also, there is no way to expand the '#var_final' default value here, just if needed,
		 * since the output could be an entirely different, e.g. if variables are used.
		 * This method also takes care of recursive '#var_final' calls
		 * (within the default value) quite well.
		 */
		$text = $varStore->mFinalizedVarsStripState->unstripGeneral( $text );
	}

	/**
	 * This will clean up the variables store after parsing has finished. It will prevent
	 * strange things to happen for example during import of several pages or job queue is running
	 * for multiple pages. In these cases variables would become some kind of superglobals,
	 * being passed from one page to the other.
	 *
	 * @param Parser $parser
	 */
	public static function onParserClearState( Parser $parser ) {
		/**
		 * MessageCaches Parser clone will mess things up if we don't reset the entire object.
		 * Only resetting the array would unset it in the original object as well! This instead
		 * will break the entire reference to the object
		 */
		$parser->mExtVariables = new self();
	}

	// Private Helper

	/**
	 * Takes care of setting a strip state pair
	 *
	 * @param string $marker
	 * @param string $value
	 */
	protected function stripStatePair( $marker, $value ) {
		$this->mFinalizedVarsStripState->addGeneral( $marker, $value );
	}

	// Public functions for interaction
	//
	// Public non-parser functions, accessible for
	// other extensions doing interactive stuff
	// with 'Variables' (like Extension:Loops)

	/**
	 * Convenience function to return the 'Variables' extensions variables store connected
	 * to a certain Parser object. Each parser has its own store which will be reset after
	 * a parsing process [Parser::parse()] has finished.
	 *
	 * @param Parser $parser
	 *
	 * @return self
	 */
	public static function get( Parser $parser ) {
		return $parser->mExtVariables;
	}

	/**
	 * Defines a variable, accessible by getVarValue() or '#var' parser function. Name and
	 * value will be trimmed and converted to string.
	 *
	 * @param string $varName
	 * @param string $value will be converted to string if no string is given
	 */
	public function setVarValue( $varName, $value = '' ) {
		$this->mVariables[ trim( $varName ) ] = trim( $value );
	}

	/**
	 * Returns a variables value or null if it doesn't exist.
	 *
	 * @param string $varName
	 * @param string|null $defaultVal
	 *
	 * @return string|null
	 */
	public function getVarValue( $varName, $defaultVal = null ) {
		$varName = trim( $varName );
		if ( $this->varExists( $varName ) ) {
			return $this->mVariables[ $varName ];
		} else {
			return $defaultVal;
		}
	}

	/**
	 * Checks whether a variable exists within the scope.
	 *
	 * @param string $varName
	 *
	 * @return bool
	 */
	public function varExists( $varName ) {
		$varName = trim( $varName );
		return array_key_exists( $varName, $this->mVariables );
	}

	/**
	 * Allows to unset a certain variable
	 *
	 * @param string $varName
	 */
	public function unsetVar( $varName ) {
		unset( $this->mVariables[ $varName ] );
	}

	/**
	 * Allows to register the usage of '#var_final'. Meaning a variable can be set as well
	 * as a default value. The return value, a strip-item then can be inserted into any
	 * wikitext processed by the same parser. Later that strip-item will be replaced with
	 * the final var text.
	 * Note: It's not possible to use the returned strip-item within other stripped text
	 *       since 'Variables' unstripping will happen before the general unstripping!
	 *
	 * @param Parser $parser
	 * @param string $varName
	 * @param string $defaultVal
	 *
	 * @return string strip-item
	 */
	public function requestFinalizedVar( Parser $parser, $varName, $defaultVal = '' ) {
		if ( $this->mFinalizedVarsStripState === null ) {
			$this->mFinalizedVarsStripState = new StripState;
		}
		$id = count( $this->mFinalizedVars );
		/*
		 * strip-item which will be unstripped in self::onInternalParseBeforeSanitize()
		 * In case the requested final variable has a value in the end, this strip-item
		 * value will be replaced with that value before unstripping.
		 */
		$rnd = "{$parser->mUniqPrefix}-finalizedvar-{$id}-" . Parser::MARKER_SUFFIX;

		$this->stripStatePair( $rnd, trim( $defaultVal ) );
		$this->mFinalizedVars[ $rnd ] = trim( $varName );

		return $rnd;
	}

}


includes/VariablesHooks.php
<?php

/**
 * Extension class containing the hook functions called by core to incorporate the
 * functionality of the Variables extension.
 */
class VariablesHooks {

	/**
	 * Sets up parser functions
	 *
	 * @since 1.4
	 *
	 * @param Parser $parser
	 */
	public static function onParserFirstCallInit( Parser $parser ) {
		/*
		 * store for variables per parser object. This will solve several bugs related to
		 * 'ParserClearState' hook clearing all variables early in combination with certain
		 * other extensions. (since v2.0)
		 */
		$parser->mExtVariables = new ExtVariables();

		self::initFunction( $parser, 'var', Parser::SFH_OBJECT_ARGS );
		self::initFunction( $parser, 'varexists', Parser::SFH_OBJECT_ARGS );
		self::initFunction( $parser, 'var_final' );
		self::initFunction( $parser, 'vardefine', Parser::SFH_OBJECT_ARGS );
		self::initFunction( $parser, 'vardefineecho', Parser::SFH_OBJECT_ARGS );
	}

	/**
	 * Does the actual setup after making sure the functions aren't disabled
	 *
	 * @param Parser $parser
	 * @param string $name The name of the parser function
	 * @param int $flags Some configuration options, see also definition in Parser.php
	 */
	private static function initFunction( Parser $parser, $name, $flags = 0 ) {
		// register function only if not disabled by configuration:
		global $egVariablesDisabledFunctions;

		if ( !in_array( $name, $egVariablesDisabledFunctions ) ) {
			$parser->setFunctionHook(
				$name,
				[ ExtVariables::class, $flags & Parser::SFH_OBJECT_ARGS ? "pfObj_$name" : "pf_$name" ],
				$flags
			);
		}
	}

	/**
	 * This will clean up the variables store after parsing has finished. It will prevent
	 * strange things to happen for example during import of several pages or job queue is running
	 * for multiple pages. In these cases variables would become some kind of superglobals,
	 * being passed from one page to the other.
	 *
	 * @param Parser $parser
	 */
	public static function onParserClearState( Parser $parser ) {
		/**
		 * MessageCaches Parser clone will mess things up if we don't reset the entire object.
		 * Only resetting the array would unset it in the original object as well! This instead
		 * will break the entire reference to the object
		 */
		$parser->mExtVariables = new ExtVariables();
	}
}


Variables_body.php
<?php
/**
 * Extension class with basic extension information. This class serves as static
 * class with the static parser functions but also als variables store instance
 * as object assigned to a Parser object.
 */
class ExtVariables {

	/**
	 * Version of the 'Variables' extension.
	 * Using this constant is deprecated, please use the data in extension.json instead.
	 * @since 1.4
	 *
	 * @var string
	 */
	const VERSION = '2.3.0';
	
	/**
	 * Internal store for variable values
	 *
	 * @private
	 * @var array
	 */
	public $mVariables = array();

	/**
	 * Array with all names of variables requested by '#var_final'. Key of the values is the
	 * stripSateId of the strip-item placed where the final var should appear.
	 *
	 * @since 2.0
	 *
	 * @private
	 * @var array
	 */
	public $mFinalizedVars = array();

	/**
	 * Variables extensions own private StripState manager to manage '#final_var' placeholders
	 * and their replacement with the final var value or a defined default.
	 *
	 * @since 2.0
	 *
	 * @private
	 * @var StripState
	 */
	public $mFinalizedVarsStripState;

	/**
	 * Sets up parser functions
	 *
	 * @since 1.4
	 */
	public static function init( Parser &$parser ) {

		/*
		 * store for variables per parser object. This will solve several bugs related to
		 * 'ParserClearState' hook clearing all variables early in combination with certain
		 * other extensions. (since v2.0)
		 */
		$parser->mExtVariables = new self();

		// Parser::SFH_OBJECT_ARGS available since MW 1.12
		self::initFunction( $parser, 'var', array( __CLASS__, 'pfObj_var' ), Parser::SFH_OBJECT_ARGS );
		self::initFunction( $parser, 'var_final' );
		self::initFunction( $parser, 'vardefine' );
		self::initFunction( $parser, 'vardefineecho' );
		self::initFunction( $parser, 'varexists' );

		return true;
	}
	private static function initFunction( Parser &$parser, $name, $functionCallback = null, $flags = 0 ) {
		if( $functionCallback === null ) {
			// prefix parser functions with 'pf_'
			$functionCallback = array( __CLASS__, 'pf_' . $name );
		}
		global $egVariablesDisabledFunctions;

		// register function only if not disabled by configuration:
		if( ! in_array( $name, $egVariablesDisabledFunctions ) ) {
			$parser->setFunctionHook( $name, $functionCallback, $flags );
		}
	}


	####################
	# Parser Functions #
	####################

	static function pf_varexists( Parser &$parser, $varName = '', $exists=true, $noexists=false ) {
		if( self::get( $parser )->varExists( $varName ) ) {
			return $exists;
		} else {
			return $noexists;
		}
	}

	static function pf_vardefine( Parser &$parser, $varName = '', $value = '' ) {
		self::get( $parser )->setVarValue( $varName, $value );
		return '';
	}

	static function pf_vardefineecho( Parser &$parser, $varName = '', $value = '' ) {
		self::get( $parser )->setVarValue( $varName, $value );
		return $value;
	}

	static function pfObj_var( Parser &$parser, $frame, $args) {
		$varName = trim( $frame->expand( $args[0] ) ); // first argument expanded already but lets do this anyway
		$varVal = self::get( $parser )->getVarValue( $varName, null );

		// default applies if var doesn't exist but also in case it is an empty string!
		if( $varVal === null || $varVal === '' ) {
			// only expand argument when needed:
			$defaultVal = isset( $args[1] ) ? trim( $frame->expand( $args[1] ) ) : '';
			return $defaultVal;
		}
		return $varVal;
	}

	static function pf_var_final( Parser &$parser, $varName, $defaultVal = '' ) {
		return self::get( $parser )->requestFinalizedVar( $parser, $varName, $defaultVal );
	}


	##############
	# Used Hooks #
	##############

	/**
	 * Used for '#var_final' parser function to insert the final variable values.
	 * @see http://www.mediawiki.org/wiki/Manual:Hooks/InternalParseBeforeSanitize
	 *
	 * @since 2.0.1
	 */
	static function onInternalParseBeforeSanitize( Parser &$parser, &$text ) {
		$varStore = self::get( $parser );

		// only do this if '#var_final' was used
		if( $varStore->mFinalizedVarsStripState === null ) {
			return true;
		}

		/*
		 * all vars are final now, check whether requested vars can be inserted for '#final_var' or
		 * if the default has to be inserted. In any case, adjust the strip item value
		 */
		foreach( $varStore->mFinalizedVars as $stripStateId => $varName ) {

			$varVal = $varStore->getVarValue( $varName, '' );
			if( $varVal !== '' ) {
				// replace strip item value with final variables value or registered default:
				$varStore->stripStatePair( $stripStateId, $varVal );
			}
		}

		/**
		 * Unstrip all '#var_final' strip-markers with their final '#var' or default values.
		 * This HAS to be done here and can't be done through the normal unstrip process of MW.
		 * This because the default value as well as the variables value stil have to be rendered properly since they
		 * may contain links or even category links. On the other hand, they can't be parsed with Parser::recursiveTagParse()
		 * since this would parse wiki templates and functions which are intended as normal text, kind of similar to
		 * returning a parser functions value with 'noparse' => true.
		 * Also, there is no way to expand the '#var_final' default value here, just if needed, since the output could be an
		 * entirely different, e.g. if variables are used.
		 * This method also takes care of recursive '#var_final' calls (within the default value) quite well.
		 */
		$text = $varStore->mFinalizedVarsStripState->unstripGeneral( $text );
		return true;
	}

	/**
	 * This will clean up the variables store after parsing has finished. It will prevent strange things to happen
	 * for example during import of several pages or job queue is running for multiple pages. In these cases variables
	 * would become some kind of superglobals, being passed from one page to the other.
	 */
	static function onParserClearState( Parser &$parser ) {
		/**
		 * MessageCaches Parser clone will mess things up if we don't reset the entire object.
		 * Only resetting the array would unset it in the original object as well! This instead
		 * will break the entire reference to the object
		 */
		$parser->mExtVariables = new self();
		return true;
	}


	##################
	# Private Helper #
	##################

	/**
	 * Takes care of setting a strip state pair
	 */
	protected function stripStatePair( $marker, $value ) {
		$this->mFinalizedVarsStripState->addGeneral( $marker, $value );
	}


	####################################
	# Public functions for interaction #
	####################################
	#
	# public non-parser functions, accessible for
	# other extensions doing interactive stuff
	# with 'Variables' (like Extension:Loops)
	#

	/**
	 * Convenience function to return the 'Variables' extensions variables store connected
	 * to a certain Parser object. Each parser has its own store which will be reset after
	 * a parsing process [Parser::parse()] has finished.
	 *
	 * @param Parser &$parser
	 *
	 * @return ExtVariables by reference so we still have the right object after 'ParserClearState'
	 */
	public static function &get( Parser &$parser ) {
		return $parser->mExtVariables;
	}

	/**
	 * Defines a variable, accessible by getVarValue() or '#var' parser function. Name and
	 * value will be trimmed and converted to string.
	 *
	 * @param string $varName
	 * @param string $value will be converted to string if no string is given
	 */
	public function setVarValue( $varName, $value = '' ) {
		$this->mVariables[ trim( $varName ) ] = trim( $value );
	}

	/**
	 * Returns a variables value or null if it doesn't exist.
	 *
	 * @param string $varName
	 * @param mixed $defaultVal
	 *
	 * @return string or mixed in case $defaultVal is being returned and not of type string
	 */
	public function getVarValue( $varName, $defaultVal = null ) {
		$varName = trim( $varName );
		if ( $this->varExists( $varName ) ) {
			return $this->mVariables[ $varName ];
		} else {
			return $defaultVal;
		}
	}

	/**
	 * Checks whether a variable exists within the scope.
	 *
	 * @param string $varName
	 *
	 * @return boolean
	 */
	public function varExists( $varName ) {
		$varName = trim( $varName );
		return array_key_exists( $varName, $this->mVariables );
	}

	/**
	 * Allows to unset a certain variable
	 *
	 * @param type $varName
	 */
	public function unsetVar( $varName ) {
		unset( $this->mVariables[ $varName ] );
	}

	/**
	 * Allows to register the usage of '#var_final'. Meaning a variable can be set as well
	 * as a default value. The return value, a strip-item then can be inserted into any
	 * wikitext processed by the same parser. Later that strip-item will be replaced with
	 * the final var text.
	 * Note: It's not possible to use the returned strip-item within other stripped text
	 *       since 'Variables' unstripping will happen before the general unstripping!
	 *
	 * @param Parser $parser
	 * @param string $varName
	 * @param string $defaultVal
	 *
	 * @return string strip-item
	 */
	function requestFinalizedVar( Parser &$parser, $varName, $defaultVal = '' ) {
		if( $this->mFinalizedVarsStripState === null ) {
			$this->mFinalizedVarsStripState = new StripState;
		}
		$id = count( $this->mFinalizedVars );
		/*
		 * strip-item which will be unstripped in self::onInternalParseBeforeSanitize()
		 * In case the requested final variable has a value in the end, this strip-item
		 * value will be replaced with that value before unstripping.
		 */
		$rnd = "{$parser->mUniqPrefix}-finalizedvar-{$id}-" . Parser::MARKER_SUFFIX;

		$this->stripStatePair( $rnd, trim( $defaultVal ) );
		$this->mFinalizedVars[ $rnd ] = trim( $varName );

		return $rnd;
	}

}
